servicetalk-grpc-netty/src/main/java/io/servicetalk/grpc/netty/DefaultGrpcClientBuilder.java

```diff
@@ -177,4 +255,57 @@ public void doAppendHttpClientFilter(final Predicate<StreamingHttpRequest> predi
                                          final StreamingHttpClientFilterFactory factory) {
         httpClientBuilder.appendClientFilter(predicate, factory);
     }
+
+    /**
+     * Extract the timeout duration from the HTTP headers if present.
+     *
+     * @param request The HTTP request to be used as source of the GRPC timeout header
+     * @return The non-negative timeout duration which may null if not present
+     * @throws IllegalArgumentException if the timeout value is malformed
+     */
+    static @Nullable Duration readTimeoutHeader(HttpRequestMetaData request) {
+        CharSequence grpcTimeoutValue = request.headers().get(GRPC_TIMEOUT_HEADER_KEY);
+        return null == grpcTimeoutValue ? null : parseTimeoutHeader(grpcTimeoutValue);
+    }
+
+    /**
+     * Parse a gRPC HTTP timeout header grpcTimeoutValue as a duration.
+     *
+     * @param grpcTimeoutValue the text value of {@link #GRPC_TIMEOUT_HEADER_KEY} header to be parsed to a timeout
+     * duration
+     * @return The non-negative timeout duration
+     * @throws IllegalArgumentException if the timeout value is malformed
+     */
+    static Duration parseTimeoutHeader(CharSequence grpcTimeoutValue) {
+        if (grpcTimeoutValue.length() < 2 || grpcTimeoutValue.length() > 9) {
+            throw new IllegalArgumentException("grpcTimeoutValue: " + grpcTimeoutValue +
+                    " (expected 2-9 characters)");
+        }
+
+        char unitChar = grpcTimeoutValue.charAt(grpcTimeoutValue.length() - 1);
+        int unitsIdx = 0;
+        while (unitsIdx < TIMEOUT_UNIT_CHARS.length && unitChar != TIMEOUT_UNIT_CHARS[unitsIdx]) {
```

Comment:
nit: `TIMEOUT_UNIT_CHARS` introduces some additional conditionals/loop that we could avoid with a `switch`:

```java
        switch (unitChar) {
            case 'n':
                units = ChronoUnit.NANOS;
                break;
            case 'u':
                units = ChronoUnit.MICROS;
                break;
                // ...
            default:
                throw new IllegalArgumentException("grpcTimeoutValue: " + grpcTimeoutValue +
                        " (Bad unit '" + unitChar + "')");
        }
```

Reply:
OK